File: client/web/divkit/tests/templates/array.test.ts
import { applyTemplate } from '../../src/utils/applyTemplate';

describe('array', () => {
    test('test_array_empty', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_empty.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_invalid_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_invalid_items.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_invalid_items_in_data', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_invalid_items_in_data.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_missing_data', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_missing_data.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_missing_one_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_missing_one_link.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_nested', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_nested.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_nested_with_internal_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_nested_with_internal_link.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_not_templated', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_not_templated.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_one_invalid_item', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_one_invalid_item.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_complex_items_with_internal_links', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_complex_items_with_internal_links.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_heterogeneous_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_heterogeneous_items.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_heterogeneous_items_with_internal_links', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_heterogeneous_items_with_internal_links.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_internal_links', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_internal_links.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_link.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_nested_templated_array', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_nested_templated_array.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_override_in_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_override_in_items.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_override_in_template', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_override_in_template.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_with_templated_items', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_with_templated_items.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_without_link', () => {
        const json = require('../../../../../test_data/template_test_data/array/test_array_without_link.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });
});

File: client/android/api-generator-test/src/test/java/com/yandex/generator/ArrayTest.kt
package com.yandex.generator

import com.yandex.div.json.ParsingException
import com.yandex.testing.Entity
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class ArrayTest {

    private val case = EntityTemplateTestCase(
        ctor = Entity.Companion::invoke
    )

    @Test
    fun `when array is defined by link it is parsed correctly`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_with_link.json"
        )
        assertEquals(ENTITY_WITH_ARRAY, actual)
    }

    @Test
    fun `when array does not contain link it is parsed correctly`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_without_link.json"
        )
        assertEquals(ENTITY_WITH_ARRAY, actual)
    }

    @Test
    fun `when array contains elements defined by links it is parsed correctly`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_with_internal_links.json"
        )
        assertEquals(ENTITY_WITH_ARRAY, actual)
    }

    @Test
    fun `when array contains links inside elements it is parsed correctly`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_with_complex_items_with_internal_links.json"
        )
        assertEquals(ENTITY_WITH_ARRAY, actual)
    }

    @Test(expected = ParsingException::class)
    fun `array has minItems 1 and empty array received`() {
        case.parse(
            directory = "array",
            filename = "test_array_empty.json"
        )
    }

    @Test(expected = ParsingException::class)
    fun `when array has all items invalid in template parsing fails`() {
        case.parse(
            directory = "array",
            filename = "test_array_invalid_items.json"
        )
    }

    @Test(expected = ParsingException::class)
    fun `when array has all items invalid in data parsing fails`() {
        case.parse(
            directory = "array",
            filename = "test_array_invalid_items_in_data.json"
        )
    }

    @Test(expected = ParsingException::class)
    fun `when data is missing parsing fails`() {
        case.parse(
            directory = "array",
            filename = "test_array_invalid_items_in_data.json"
        )
    }

    @Test
    fun `when array has some invalid items they are skipped`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_one_invalid_item.json"
        )
        assertEquals(ENTITY_WITH_ARRAY, actual)
    }

    @Test
    fun `when array has some invalid links they are skipped`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_missing_one_link.json"
        )
        assertEquals(ENTITY_WITH_ARRAY, actual)
    }

    @Test
    fun `when array is not templated parsing succeeds`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_not_templated.json"
        )
        assertEquals(ENTITY_WITH_ARRAY, actual)
    }

    @Test
    fun `when array has heterogeneous items parsing succeeds`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_with_heterogeneous_items.json"
        )
        assertEquals(ENTITY_WITH_HETEROGENOUS_ARRAY, actual)
    }

    @Test
    fun `when array has heterogeneous items with internal links parsing succeeds`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_with_heterogeneous_items_with_internal_links.json"
        )
        assertEquals(ENTITY_WITH_HETEROGENOUS_ARRAY, actual)
    }

    @Test
    fun `when array has nested array parsing succeeds`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_nested.json"
        )
        assertEquals(ENTITY_WITH_NESTED_ARRAY, actual)
    }

    @Test
    fun `when array has nested array with links inside parsing succeeds`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_nested_with_internal_link.json"
        )
        assertEquals(ENTITY_WITH_NESTED_ARRAY, actual)
    }

    @Test
    fun `when array has nested templated array parsing succeeds`() {
        val actual = case.parse(
            directory = "array",
            filename = "test_array_nested_with_internal_link.json"
        )
        assertEquals(ENTITY_WITH_NESTED_ARRAY, actual)
    }
}
